core/src/test/java/feign/FeignTest.java

```diff
@@ -388,6 +388,23 @@ public Object decode(Response response, Type type) throws IOException {
   }
 
   @Test
+  public void decoderCanThrowUnwrappedExceptionInDecode404Mode() throws Exception {
```

Comment:
let's make this more of a story.

`decoderCanThrowUnwrappedExceptionInDecode404Mode` doesn't tell what we want, it tells us what we don't want.
how about `decode404AllowsCustomExceptions`

Instead of curiously returning "success!" to demonstrate 404, return 404?

Then, in the decode method, assert response.code is 404 (to reinforce your point). Then, throw an IllegalStateException or NoSuchElementException, which is more like what you're looking for.

When a test is written like this, future me's will surely delete it, because it makes no sense to have a feature like this. When a test show's the intent, it is less likely to be deleted or abandoned in a future version. Basically, think of your test as an absentee advocate for your feature!